src/visitors/hoist.js

```diff
@@ -183,122 +182,201 @@ module.exports = {
+    for (let specifier of path.node.specifiers) {
+      if (t.isImportDefaultSpecifier(specifier)) {
+        path.scope.rename(
+          specifier.local.name,
+          getName(asset, 'import', path.node.source.value, 'default')
         );
-        path.remove();
-      } else {
-        path.replaceWith(
-          t.variableDeclaration(
-            'var',
-            specifiers.map(specifier =>
-              t.variableDeclarator(
-                specifier.local,
-                t.memberExpression(
-                  t.identifier(
-                    '$' + asset.id + '$require$' + t.toIdentifier(source.value)
-                  ),
-                  specifier.imported
-                )
-              )
-            )
+      } else if (t.isImportSpecifier(specifier)) {
+        path.scope.rename(
+          specifier.local.name,
+          getName(
+            asset,
+            'import',
+            path.node.source.value,
+            specifier.imported.name
           )
         );
+      } else if (t.isImportNamespaceSpecifier(specifier)) {
+        path.scope.rename(
+          specifier.local.name,
+          getName(asset, 'require', path.node.source.value)
+        );
       }
     }
+
+    path.remove();
   },
-  ExportNamedDeclaration(path, asset) {
-    let {declaration, source, specifiers} = path.node;
 
-    if (!source) {
-      let declarations;
+  ExportDefaultDeclaration(path, asset) {
+    let {declaration} = path.node;
+    let identifier = getIdentifier(asset, 'export', 'default');
 
-      if (declaration) {
-        declarations = declaration.declarations.map(decl => {
-          asset.cacheData.exports.push(decl.id.name);
+    if (t.isExpression(declaration)) {
+      // Declare a variable to hold the exported value.
+      path.replaceWith(
+        t.variableDeclaration('var', [
+          t.variableDeclarator(identifier, declaration)
+        ])
+      );
+    } else {
+      // Rename the declaration to the exported name.
+      path.replaceWith(declaration);
+      path.scope.rename(declaration.id.name, identifier.name);
+    }
 
-          return getNamedExportVarDecl(asset, decl.id, decl.init);
-        });
-      } else if (specifiers.length > 0) {
-        declarations = t.variableDeclaration(
-          'var',
-          specifiers.map(specifier => {
-            asset.cacheData.exports.push(specifier.exported.name);
+    // Add assignment to exports object for namespace imports and commonjs.
+    path.insertAfter(
+      EXPORT_ASSIGN_TEMPLATE({
+        EXPORTS: getExportsIdentifier(asset),
+        NAME: t.identifier('default'),
+        LOCAL: identifier
+      })
+    );
+
+    asset.cacheData.exports[identifier.name] = 'default';
+
+    // Mark the asset as an ES6 module, so we handle imports correctly in the packager.
+    asset.cacheData.isES6Module = true;
+  },
+
+  ExportNamedDeclaration(path, asset) {
+    let {declaration, source, specifiers} = path.node;
+
+    if (source) {
```

Comment:
Support re-exports, e.g. `export {foo} from 'bar'`. Again, this will be problematic I realized, and the code to insert a variable here probably won't work right.